core/storage/base_model/gae_models.py

```diff
@@ -105,6 +105,27 @@ def __init__(self, *args, **kwargs):
         super(BaseModel, self).__init__(*args, **kwargs)
         self._last_updated_timestamp_is_fresh = False
 
+    def clone(self, **new_values):
```

Comment:
I am a bit concerned about this, it is a fundamental infrastructure change. Why do we need to do this at all?

/cc @vojtechjelinek

Reply:
We need this because all inputs provided by the beam infra must be treated as immutable.

NDB models are not immutable, hence the necessity to clone them before performing any operation on them.

FWIW, validation operations will not mutate any models anyway because they're read-only by nature. However, **I'd prefer every operation clone a model (guaranteeing immutability), than some jobs forgetting to clone a model and leading to hard-to-understand bugs (especially for new contributors).**

Reply:
Why must all inputs provided by the beam infra be treated as immutable? I don't understand the difference between Beam and how we currently run our jobs in that respect.

Some existing jobs take a model, mutate it, and then put() it. How would beam handle that?

Reply:
Because that is Apache Beam's contract. The infra is structured such that a model instance can be passed to many distinct operations (e.g., it caches values). Having one of those operations modify the model is undefined behavior.

If an operation needs to mutate and put it, that's fine. Just clone it first. However, I'd rather have every single operation clone than only some of them.

Reply:
I see, OK. But I don't want this to be part of the base model infrastructure API, since we should not have multiple model instances with the same ID floating around the codebase, any of which can be "put()" at any time. Can it be written in such a way that it is done by default for Apache Beam jobs only?

Reply:
That sounds fair, I think it should be possible by having job authors inherit from our own base class, rather than from `beam.DoFn`.

Then, we can move this code into that base class. /cc @shubha-rajan

Reply:
Done!